File: backend/src/infrastructure/repositories/candidate.repository.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { ICandidateRepository } from '@/domain/repositories/candidate.repository.interface';

@Injectable()
export class CandidateRepository implements ICandidateRepository {
  constructor(private prisma: PrismaService) {}

  async findAll(filters: any) {
    const {
      page = 1,
      pageSize = 20,
      keyword,
      positionId,
      department,
      stage,
      source,
      ownerId,
      createdFrom,
      createdTo,
    } = filters;

    const where: any = {};

    if (keyword) {
      where.OR = [
        { name: { contains: keyword } },
        { phone: { contains: keyword } },
        { email: { contains: keyword } },
        { currentCompany: { contains: keyword } },
      ];
    }
    if (positionId) where.positionId = positionId;
    if (stage) where.stage = stage;
    if (source) where.source = source;
    if (ownerId) where.ownerId = ownerId;
    if (department) where.position = { department };
    if (createdFrom || createdTo) {
      where.createdAt = {};
      if (createdFrom) where.createdAt.gte = new Date(createdFrom);
      if (createdTo) where.createdAt.lte = new Date(createdTo);
    }

    const [data, total] = await Promise.all([
      this.prisma.candidate.findMany({
        where,
        skip: (page - 1) * pageSize,
        take: pageSize,
        include: {
          position: { select: { id: true, title: true, department: true } },
          owner: { select: { id: true, name: true } },
          tags: true,
          files: { where: { isResume: true }, take: 1 },
        },
        orderBy: { updatedAt: 'desc' },
      }),
      this.prisma.candidate.count({ where }),
    ]);

    return { data, total };
  }

  async findById(id: string) {
    return this.prisma.candidate.findUnique({
      where: { id },
      include: {
        position: true,
        owner: { select: { id: true, name: true, email: true } },
        tags: true,
        files: true,
        interviews: {
          include: {
            interviewer: { select: { id: true, name: true } },
            feishuCalendarEvent: true,
          },
          orderBy: { scheduledAt: 'desc' },
        },
        interviewFeedbacks: {
          include: { evaluator: { select: { id: true, name: true } } },
          orderBy: { createdAt: 'desc' },
        },
        offerApprovals: { orderBy: { createdAt: 'desc' } },
      },
    });
  }

  async create(data: any) {
    const { tags, ...rest } = data;
    return this.prisma.candidate.create({
      data: {
        ...rest,
        tags: tags?.length
          ? { create: tags.map((tag: string) => ({ tag })) }
          : undefined,
      },
      include: {
        position: true,
        owner: { select: { id: true, name: true } },
        tags: true,
      },
    });
  }

  async update(id: string, data: any) {
    const { tags, ...rest } = data;
    if (tags) {
      await this.prisma.candidateTag.deleteMany({ where: { candidateId: id } });
    }
    return this.prisma.candidate.update({
      where: { id },
      data: {
        ...rest,
        tags: tags?.length
          ? { create: tags.map((tag: string) => ({ tag })) }
          : undefined,
      },
      include: {
        position: true,
        owner: { select: { id: true, name: true } },
        tags: true,
      },
    });
  }

  async delete(id: string) {
    await this.prisma.candidate.delete({ where: { id } });
  }

  async batchUpdate(ids: string[], data: any) {
    const result = await this.prisma.candidate.updateMany({
      where: { id: { in: ids } },
      data,
    });
    return result.count;
  }
}

File: backend/src/main/java/com/company/hr/repository/CandidateRepository.java
package com.company.hr.repository;

import com.company.hr.entity.Candidate;
import com.company.hr.enums.CandidateStage;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.time.LocalDateTime;
import java.util.List;

@Repository
public interface CandidateRepository extends JpaRepository<Candidate, String> {

    List<Candidate> findByStage(CandidateStage stage);

    @Query("SELECT c FROM Candidate c " +
           "WHERE (:keyword IS NULL OR c.name LIKE %:keyword% OR c.phone LIKE %:keyword% " +
           "OR c.email LIKE %:keyword% OR c.currentCompany LIKE %:keyword%) " +
           "AND (:positionId IS NULL OR c.position.id = :positionId) " +
           "AND (:stage IS NULL OR c.stage = :stage) " +
           "AND (:source IS NULL OR c.source = :source) " +
           "AND (:ownerId IS NULL OR c.owner.id = :ownerId) " +
           "AND (:department IS NULL OR c.position.department = :department) " +
           "AND (:createdFrom IS NULL OR c.createdAt >= :createdFrom) " +
           "AND (:createdTo IS NULL OR c.createdAt <= :createdTo)")
    Page<Candidate> findByFilters(@Param("keyword") String keyword,
                                  @Param("positionId") String positionId,
                                  @Param("stage") CandidateStage stage,
                                  @Param("source") com.company.hr.enums.CandidateSource source,
                                  @Param("ownerId") String ownerId,
                                  @Param("department") String department,
                                  @Param("createdFrom") LocalDateTime createdFrom,
                                  @Param("createdTo") LocalDateTime createdTo,
                                  Pageable pageable);

    @Modifying
    @Query("UPDATE Candidate c SET c.stage = :stage WHERE c.id IN :ids")
    int batchUpdateStage(@Param("ids") List<String> ids, @Param("stage") CandidateStage stage);

    @Modifying
    @Query("UPDATE Candidate c SET c.owner.id = :ownerId WHERE c.id IN :ids")
    int batchUpdateOwner(@Param("ids") List<String> ids, @Param("ownerId") String ownerId);
}
